Dev/explore edit

RandoomWalks opened the pull request:
Comprehensive set of code changes to address the concurrency issues in the KademliaNode structure and its associated methods

RandoomWalks committed:
improve error handling consistency

RandoomWalks committed:
Exponential Backoff: Introduced exponential backoff with retries for handling timeouts in network operations.

Retry Logic: Integrated the retry logic in critical methods to improve the reliability of the Kademlia node, making it more resilient to transient network issues.

Configuration Flexibility: The key and value size limits are now configurable through the Config struct, allowing the application to be tailored to specific needs.
Validation Updates: The validate_key_value method uses these configurable limits instead of hardcoded values, improving flexibility and adaptability.

Use of RwLock: Introduced tokio::sync::RwLock to wrap the DashMap in StorageManager, allowing safe concurrent reads and writes.
Proper Synchronization: Ensured that all access to shared resources is synchronized, preventing race conditions.

RandoomWalks committed:
Increased Robustness: By retrying the bootstrap process after a delay, the node can handle transient network issues more gracefully and reduce the risk of prematurely assuming it is the first node in the network.

Prevention of Network Partitioning: This strategy helps prevent scenarios where multiple nodes incorrectly believe they are the first in the network, which could lead to network partitioning or split-brain conditions.

RandoomWalks committed:
Node Bootstrap- Upgrades made
 add a maximum retry limit
   adjust the jitter calculation to take into account the retry count,
  add a minimum successful contacts requirement before proceeding with the bootstrap process
  backoff strategy for the retry_delay itself, increasing it gradually after each failed loop iteration.
 introduce a randomized startup delay before the first bootstrap attempt to spread out the load.

hwrdtm commented on src/node.rs:
```diff
@@ -709,32 +774,41 @@ impl KademliaNode {
+                    )
+                    .await
+                    {
+                        Ok(Ok((Message::Pong { sender }, src))) => {
```
nested matching like so is not as legible as just doing separate `match` statements.

hwrdtm commented on src/node.rs:
```diff
@@ -177,176 +298,118 @@ impl KademliaNode {
+                                .iter()
+                                .any(|(_, addr)| *addr == contact)
+                            {
+                                self.routing_table.update(self.id, contact);
```
why do we need to place this inside the if block? Just populate anyways?

hwrdtm commented on src/node.rs:
```diff
@@ -36,46 +65,91 @@ impl NetworkManager {
 }
 
 // Storage Manager
 pub struct StorageManager {
```
This should be a trait so we can have in-memory storage or other persistent storage implementations to swap out.

RandoomWalks committed:
Refactor NetworkManager to improve ownership handling and resolve lifetime errors

- Added `Clone` trait implementation for `NetworkManager` to allow cloning instances for use across asynchronous tasks.
- Refactored `send_message` and `receive_message` methods to use owned `Message` instances, avoiding references and reducing lifetime complexities.
- Introduced `Arc<Message>` wrapping to manage shared ownership of messages safely across multiple closures and async blocks.
- Corrected `Arc` cloning inside the `exponential_backoff` closure to prevent moving captured variables, ensuring proper use of shared ownership patterns.
- Updated method calls to `send_message` to pass `Message` by value rather than reference, aligning with the new function signature and improving clarity.
- Refined exponential backoff retry logic by cloning required `Arc` objects inside closures, preserving the original references.
- Enhanced error handling in various methods to accommodate changes in ownership and message passing, ensuring robustness in handling network operations.
